### DISCUSSION TITLE
Why do I have to set properties that have defaults when inserting into a tabel/model?

### DISCUSSION THREAD

**Discussion by turbo-turnip** [prisma/prisma#11729] [Q&A]
Hello,
I have this model in my schema:
![Screen Shot 2022-02-09 at 6 49 17 PM](https://user-images.githubusercontent.com/75766911/153327689-4a61cd08-73f4-44b9-ad46-d9a9a56e05d5.png)
It has defaulted fields `details`, `avatar`, `likes`, and `activity`.
When I try to create a user, I use this in my backend (TypeScript):
![Screen Shot 2022-02-09 at 6 50 49 PM](https://user-images.githubusercontent.com/75766911/153327849-068f4b2e-127b-40db-bc91-930c8b387a17.png)
However, I get this error:
`Type '{ name: string; email: string; password: string; }' is not assignable to type '(Without<UserCreateInput, UserUncheckedCreateInput> & UserUncheckedCreateInput) | (Without<...> & UserCreateInput)'.`
I have found that when I set the `details`, `avatar`, `likes`, and `activity` fields, the error goes away.
Why can't I just set the `name`, `email`, and `password` fields? Shouldn't I just be able to set those and not have to worry about the others since they have default values?

It could be the way how I migrate my database though, here's how I do it: 
npx prisma migrate reset  
npx prisma generate (I use reset since I am just testing out inserting into the database for now)

Node: v16.6.2
prisma: 3.9.1
@prisma/client: 3.9.1
nodemon: 2.0.15
ts-node: 10.4.0

**Comment by nurul3101** [comment]
Hey @SoftwareFuze 👋🏻 ,

In your `User` model you have kept `details`, `avatar`, `likes` and `activity` as required, that's the reason you are getting the typescript error.

Adding `?` to the fields which are optional should solve your issue.

```prisma
model User {
  details  String?
  avatar   String?
  likes    Int?
  activity String?
}
```
Reference to [Optional and Mandatory fields](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#optional-and-mandatory-fields) in Prisma Schema.

Please let me know if this solves it for you or in case you encounter any other issues.

**Reply by Eddie2111** [reply]
##############################################

Guys, if any of you still facing the same issue, the reply above this comment is the solution.

Check your schema, you are missing one of the required fields and that's what creating the issue

##############################################

**Reply by brunobely** [reply]
If a field has a `@default` it is not required (for insertion), though?

**Comment by richardwardza** [comment]
I have 
```
model Abc {
  id           String           @id @default(cuid())
  createdAt    DateTime         @default(now())
  updatedAt    DateTime         @default(now())
  version      Int              @default(value: 1)
   name      String
}
```
and can insert using:
```
 prisma.abc.create({data: { name: "hello world"}});
 ```      
 
 @SoftwareFuze  what is the full error you are receiving?
 Have you run a `yarn prisma generate` lately to ensure you have the correct updated types?

**Reply by vladi-stevanovic** [reply]
👋 Hello @SoftwareFuze  I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! 🙌

**Comment by primostasis123** [comment]
i have this problem i dont get it

**Reply by nurul3101** [reply]
Can you please open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/new/choose) so that we can discuss the issue you are facing?

**Comment by devpilot-official** [comment]
Hello, Please, has anyone been able to solve this problem I am having the same issue? Thanks

**Reply by amakavanessa** [reply]
Hello, I encountered a similar issue where I needed to resolve an error with the data object. To address this, I made sure to include all the required fields on my schema in the this.prisma.user.create() function. Additionally, since my DTO was of type AuthDto, I incorporated the required fields within the class itself. This approach allowed me to easily access the fields. if after all the problem persists try restarting your code editor.
![Screenshot 2023-07-13 135557](https://github.com/prisma/prisma/assets/54083444/00b6f929-eefe-4c43-9d7e-7fa63ed9fa31)
![Screenshot 2023-07-13 135533](https://github.com/prisma/prisma/assets/54083444/49e3be3c-16f8-49e8-baf5-e93e2100e3d4)
![Screenshot 2023-07-13 135613](https://github.com/prisma/prisma/assets/54083444/af79c7b0-278f-4fac-ac88-ba28551a2084)

**Comment by Coderamrin** [comment]
So, whenever you have required fields on the schema add those fields. And if you update the Prisma Schema run this command `npx prisma generate`

**Comment by rajuproject** [comment]
run this command in your vs code   npx prisma generate

**Comment by Aakash-Yadav23** [comment]
i also had same issue .
 
just do this in your controller  you are importing prismaclient  

import Prisma also like this 
![Screenshot from 2024-01-24 04-01-34](https://github.com/prisma/prisma/assets/110857134/b31e62ba-c5b7-45b8-846a-a7229b81da1a)

then in your controller do like this 

![Screenshot from 2024-01-24 04-03-01](https://github.com/prisma/prisma/assets/110857134/1984747c-24a4-4975-8df9-39887c3a4a1b)

**Comment by FaizaanFazal** [comment]
For Default values required by typescript you can use:
` Omit<model_type_here, 'id'>[];` // list: as my data for model was comming in list

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by Portur** [comment]
I am running into the same problem but no one here is explaining the issue.
Here is what is going on.

If you have model such as this:
```prisma
        model myTable {
          // columns
          id      String  @unique
          name    String  @db.VarChar(100)
          surname String? @db.VarChar(100)
          speed      String    @default("fast")
          ignore  Boolean @default(false)
        
          // defaults
          createdAt DateTime @default(now()) @map("created_at")
          updatedAt DateTime @updatedAt @map("updated_at")
        
          @@index([id])
          @@map("my_table")
          @@schema("my_schema")
        }
```

The expectation here is I want to be able to CREATE or UPDATE the table without the need of sending values for columns that have `@default(...)` BUT at the same time I DO NOT want them to be NULLABLE.

So in this case I want to do

```typescript
      await this.myTable.upsert({
            update: {
              name: 'new name'
            },
            create: {
              speed: 'slow',
              name: 'my name',
            },
            where: {
              id: '123'
            }
          })
```

In the above case I am not sending data for `createdAt`, `updateAt` and `speed` since they have defaults values, however they should not be null. This is where the issue is. I can understand I have set `surname` to optional and thus it is nullable, but I don't want the others nullable.

So where the issue comes into exactly is when using the type `MyTable` it throws a TS error which is mentioned in above posts.

ie.

```typescript
import { Prisma } from "@prisma/client";

const user: MyTable = {
   name: 'new name',
   id : '123'
}

await upsertMyTable(user)



async function upsertMyTable(user: MyTable): MyTable {
  await this.myTable.upsert({
      update: user,
      create: user,
      where: {
        id: user.id
      }
    })
}
```

The question then here is which type should be used for the input param of `function upsertMyTable(user: <which type to use here>): MyTable { ... }`

Should it be `MyTableUncheckedCreateInput` or `MyTableUpdateInput` or ... ?
